### PR TITLE
fix: Headless/unattended install hangs in embedded clusters with recent kubernetes versions

### PR DESCRIPTION
**Pull request by sgalsaleh**
#### What this PR does / why we need it:

Fixes an issue where headless/unattended installs hang in embedded clusters with recent Kubernetes versions.

#### Which issue(s) this PR fixes:

Fixes [SC-55766](https://app.shortcut.com/replicated/story/55766/headless-unattended-install-does-not-work-in-kurl-clusters)

#### Does this PR introduce a user-facing change?

```release-note
Fixes an issue where headless/unattended installs hang in embedded clusters with recent Kubernetes versions.
```

#### Does this PR require documentation?

NONE

### CODE REVIEW

File: pkg/k8sutil/portforward.go [hunk 1]
```diff
@@ -26,39 +27,90 @@ import (
...
+		return 0, errors.Wrap(err, "failed to get free port")
+	}
+
+	isPortAvailable, err := IsPortAvailable(clientset, port)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check if port is available")
+	}
+
+	if !isPortAvailable {
+		return FindFreePort(clientset)
```
Comment (divolgin) [inline pkg/k8sutil/portforward.go:77]: Wanna put a limit on this recursion?  If `isPortAvailable` stays false, this will spin forever.

Comment (sgalsaleh) [reply pkg/k8sutil/portforward.go:77]: yeah, i knew you were gonna say that... okay :)

File: pkg/k8sutil/portforward.go [hunk 2]
```diff
@@ -26,39 +27,90 @@ import (
...
 	host := net.JoinHostPort("127.0.0.1", strconv.Itoa(port))
 	server, err := net.Listen("tcp4", host)
 	if err != nil {
-		return false
+		return false, nil
 	}
 
 	server.Close()
-	return true
+	return true, nil
```
Comment (divolgin) [inline pkg/k8sutil/portforward.go:62]: Should we still do a net.Listen check at this point?  This function does not account for ports being in use by system processes that run outside of cluster.

Comment (sgalsaleh) [reply pkg/k8sutil/portforward.go:62]: it accounts for ports being used where the kots commands are being run though, and that's still useful?

Comment (divolgin) [reply pkg/k8sutil/portforward.go:62]: yeah, nm, the code i had in mind is already in this function

**Review by divolgin — DISMISSED** [review]

**Review by divolgin — APPROVED** [review]

**Pull request merged**